clean.cmd

```diff
@@ -0,0 +1,177 @@
+  set clean_bin=true
+  set clean_pgk=true
+  set clean_pgkcache=true
+  set clean_all=true
+  set thisArgs=!thisArgs!%1
+  goto Next
+)
+
+if [!thisArgs!]==[] (
+  set unprocessedBuildArgs=!allargs!
+) else (
+  call set unprocessedBuildArgs=%%allargs:*!thisArgs!=%%
+)
+
+:Next
+shift /1
+goto Loop
+
+:Begin
+if /I [%clean_environment%] == [true] (
+  call :CleanEnvironment
+)
+
+if /I [%clean_src%] == [true] (
+  echo Cleaning src directory ...
+  echo. >> %cleanlog% && echo git clean -xdf %~dp0src >> %cleanlog%
+  call git clean -xdf %~dp0src >> %cleanlog%
+  call :CheckErrorLevel
+)
+
+if /I [%clean_bin%] == [true] (
+  echo Cleaning bin directory ...
+  echo. >> %cleanlog% && echo %~dp0executor.cmd !unprocessedBuildArgs! clean -b>> %cleanlog%
+  call %~dp0executor.cmd !unprocessedBuildArgs! clean -b>> %cleanlog%
+  call :CheckErrorLevel
+)
+
+if /I [%clean_pgk%] == [true] (
+  echo Cleaning package directory ...
+  echo. >> %cleanlog% && echo %~dp0executor.cmd !unprocessedBuildArgs! clean -p>> %cleanlog%
+  call %~dp0executor.cmd !unprocessedBuildArgs! clean -p >> %cleanlog%
+  call :CheckErrorLevel
+)
+
+if /I [%clean_pgkcache%] == [true] (
+  echo Cleaning package cache directory ...
+  echo. >> %cleanlog% && echo %~dp0executor.cmd !unprocessedBuildArgs! clean -c >> %cleanlog%
+  call %~dp0executor.cmd !unprocessedBuildArgs! clean -c >> %cleanlog%
+  call :CheckErrorLevel
+)
+
+if /I [%clean_tools%] == [true] (
+  call :CleanEnvironment
+  echo Cleaning tools directory ...
+  echo. >> %cleanlog% && echo rmdir /s /q %~dp0tools >> %cleanlog%
+  rmdir /s /q %~dp0tools >> %cleanlog%
+  REM Don't call CheckErrorLevel because if the Tools directory didn't exist when this script was
+  REM invoked, then it sometimes exits with error level 3 despite successfully deleting the directory.
+)
+
+if /I [%clean_all%] == [true] (
+  call :CleanEnvironment
+  echo Cleaning entire working directory ...
+  echo. >> %cleanlog% && echo git clean -xdf -e clean.log %~dp0 >> %cleanlog%
+  call git clean -xdf -e clean.log %~dp0 >> %cleanlog%
+  call :CheckErrorLevel
+)
+
+if /I [%clean_successful%] == [true] (
+  echo Clean completed successfully.
+  echo. >> %cleanlog% && echo Clean completed successfully. >> %cleanlog%
+  exit /b 0
+) else (
+  echo An error occured while cleaning; see %cleanlog% for more details.
+  echo. >> %cleanlog% && echo Clean completed with errors. >> %cleanlog%
+  exit /b 1
+)
+
+:Usage
+echo.
+echo Repository cleaning script.
+echo.
+echo Options:
+echo     -b     - Deletes the binary output directory.
+echo     -p     - Deletes the repo-local nuget package directory.
+echo     -c     - Deletes the user-local nuget package cache.
+echo     -t     - Deletes the tools directory.
+echo     -s     - Deletes the untracked files under src directory (git clean src -xdf).
+echo     -e     - Kills and/or stops the processes that are still running, for example VBCSCompiler.exe
+echo     -all   - Combines all of the above.
```

Comment:
At some point I would like to see if we can clean up the clean script such that everything but "all" will just delegate to the build engine and the all will delegate then clean tools directory as a follow-up.  This can happen later.